ubuntu_image/parser.py

```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
-            # is only allowed for GPT, while 2-digit-only is only allowed for
-            # MBR.  Note too that 2-item tuples are also already ensured.
+            structure_role = structure.get('role')
+            is_mbr = \
+                structure_type == 'mbr' or structure_role is StructureRole.mbr
+            # Validate structure types and roles. These can be either GUIDs,
+            # two hex digits or hybrids.  The basic syntactic validation
+            # happens above in the Voluptuous schema, but here we need to
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
+                    raise GadgetSpecificationError(
+                        'Type mbr and role fields assigned at the same time, '
+                        'please use role field only')
             if (isinstance(structure_type, UUID) and
                     schema is not VolumeSchema.gpt):
                 raise GadgetSpecificationError(
-                    'MBR structure type with non-MBR schema')
+                    'GUID structure type with non-GPT schema')
             elif (isinstance(structure_type, str) and
-                    structure_type != 'mbr' and
+                    not is_mbr and
                     schema is not VolumeSchema.mbr):
                 raise GadgetSpecificationError(
-                    'GUID structure type with non-GPT schema')
+                    'MBR structure type with non-MBR schema')
             # Check for implicit vs. explicit partition offset.
             if offset is None:
                 # XXX: Ensure the special case of the 'mbr' type doesn't
                 # extend beyond the confines of the mbr.
-                if structure_type != 'mbr' and last_offset < MiB(1):
+                if (structure_role is not StructureRole.mbr and
+                        structure_type is not 'mbr' and
+                        last_offset < MiB(1)):
                     offset = MiB(1)
                 else:
                     offset = last_offset
             last_offset = offset + size
             # Extract the rest of the structure data.
             structure_id = structure.get('id')
             filesystem = structure['filesystem']
-            if structure_type == 'mbr':
-                if structure_id is not None:
+            if structure_role is StructureRole.mbr:
```

Comment:
Yes, if you see previous commits this is how it was done originally, this was the 'nice' solution I did. This has been changed as we thought that there are differences between how the legacy 'mbr' type was handled and the new mbr role - which, as per the mailing thread, has been resolved.